components/feature/addons/src/main/java/mozilla/components/feature/addons/migration/SupportedAddonsChecker.kt

```diff
@@ -0,0 +1,204 @@
+        WorkManager.getInstance(applicationContext)
+                .cancelUniqueWork(CHECKER_UNIQUE_PERIODIC_WORK_NAME)
+        logger.info("Unregister check for new supported add-ons")
+    }
+
+    @VisibleForTesting
+    internal fun createPeriodicWorkerRequest(): PeriodicWorkRequest {
+        return PeriodicWorkRequestBuilder<SupportedAddonsWorker>(
+                frequency.repeatInterval,
+                frequency.repeatIntervalTimeUnit
+        ).apply {
+            setConstraints(getWorkerConstrains())
+            addTag(WORK_TAG_PERIODIC)
+        }.build()
+    }
+
+    @VisibleForTesting
+    internal fun getWorkerConstrains() = Constraints.Builder()
+            .setRequiredNetworkType(NetworkType.CONNECTED)
+            .build()
+
+    companion object {
+        private const val IDENTIFIER_PREFIX = "mozilla.components.feature.addons.migration"
+
+        @VisibleForTesting
+        internal const val CHECKER_UNIQUE_PERIODIC_WORK_NAME =
+                "$IDENTIFIER_PREFIX.DefaultSupportedAddonsChecker.periodicWork"
+
+        /**
+         * Identifies all the workers that periodically check for new add-ons.
+         */
+        @VisibleForTesting
+        internal const val WORK_TAG_PERIODIC =
+                "$IDENTIFIER_PREFIX.DefaultSupportedAddonsChecker.periodicWork"
+
+        internal fun createDefaultNotificationIntent(content: Context): Intent {
+            return content.packageManager.getLaunchIntentForPackage(content.packageName)?.apply {
+                flags = Intent.FLAG_ACTIVITY_NEW_TASK
+            } ?: throw IllegalStateException("Package has no launcher intent")
+        }
+    }
+}
+
+/**
+ * A implementation which uses WorkManager APIs to check for new available supported add-ons.
+ */
+internal class SupportedAddonsWorker(
+    val context: Context,
+    params: WorkerParameters
+) : CoroutineWorker(context, params) {
+    private val logger = Logger("SupportedAddonsWorker")
+
+    @Suppress("TooGenericExceptionCaught", "MaxLineLength")
+    override suspend fun doWork(): Result {
+
+        return try {
+            logger.info("Trying to check for new supported add-ons")
+
+            val addonManager = GlobalAddonDependencyProvider.requireAddonManager()
+
+            val newSupportedAddons = addonManager.getAddons().filter { addon ->
+                addon.isSupported() && addon.isDisabledAsUnsupported()
+            }
+
+            if (newSupportedAddons.isNotEmpty()) {
+                val extIds = newSupportedAddons.joinToString { it.id }
+                logger.info("New supported add-ons available $extIds")
+                createNotification()
+            }
+            Result.success()
+        } catch (exception: Exception) {
+            logger.error("An exception happened trying to check for new supported add-ons, re-schedule ${exception.message}", exception)
+            GlobalAddonDependencyProvider.onCrash?.invoke(exception)
+            Result.success()
```

Comment:
`Result.failure()`?

Reply:
It's a bit intuitive but as we are using a [periodic work request](https://developer.android.com/reference/androidx/work/WorkManager#enqueueUniquePeriodicWork(java.lang.String,%20androidx.work.ExistingPeriodicWorkPolicy,%20androidx.work.PeriodicWorkRequest)), if use[ Result.failure()](https://developer.android.com/reference/androidx/work/ListenableWorker.Result?hl=en#failure() ), we're going to prevent the work from happening again  "failure indicates a permanent stoppage of the chain of work". As we didn't defined which types of exceptions must stop the checker I prefer to pass the crash to the crash reporter and let the work run again until we found the cause.